dll/win32/browseui/explorerband.cpp

```diff
@@ -457,28 +506,55 @@ LRESULT CExplorerBand::OnContextMenu(UINT uMsg, WPARAM wParam, LPARAM lParam, BO
 
     IUnknown_SetSite(contextMenu, (IDeskBand *)this);
 
+    if (SUCCEEDED(pFolder->GetAttributesOf(1, (LPCITEMIDLIST*)&pidlChild, &attr)) && (attr & SFGAO_CANRENAME))
+        cmf |= CMF_CANRENAME;
+
     treeMenu = CreatePopupMenu();
-    hr = contextMenu->QueryContextMenu(treeMenu, 0, FCIDM_SHVIEWFIRST, FCIDM_SHVIEWLAST,
-        CMF_EXPLORE);
+    cmdBase = max(FCIDM_SHVIEWFIRST, 1);
+    hr = contextMenu->QueryContextMenu(treeMenu, 0, cmdBase, FCIDM_SHVIEWLAST, cmf);
     if (!SUCCEEDED(hr))
     {
         WARN("Can't get context menu for item\n");
         DestroyMenu(treeMenu);
         goto Cleanup;
     }
+
     uCommand = TrackPopupMenu(treeMenu, TPM_LEFTALIGN | TPM_RETURNCMD | TPM_LEFTBUTTON | TPM_RIGHTBUTTON,
-        x, y, 0, m_hWnd, NULL);
+        pt.x, pt.y, 0, m_hWnd, NULL);
+    if (uCommand)
+    {
+        uCommand -= cmdBase;
 
-    ExecuteCommand(contextMenu, uCommand);
+        // Do DFM_CMD_RENAME in the treeview
+        if ((cmf & CMF_CANRENAME) && SHELL_IsVerb(contextMenu, uCommand, L"rename"))
+        {
+            HTREEITEM oldSelected = m_oldSelected;
+            SetFocus();
+            startedRename = TreeView_EditLabel(m_hWnd, item) != NULL;
+            m_oldSelected = oldSelected; // Restore after TVN_BEGINLABELEDIT
+            goto Cleanup;
+        }
+
+        hr = ExecuteCommand(contextMenu, uCommand);
+    }
 
 Cleanup:
     if (contextMenu)
         IUnknown_SetSite(contextMenu, NULL);
     if (treeMenu)
         DestroyMenu(treeMenu);
-    m_bNavigating = TRUE;
-    TreeView_SelectItem(m_hWnd, m_oldSelected);
-    m_bNavigating = FALSE;
+    if (startedRename)
+    {
+        // The treeview disables drawing of the edited item so we must make sure
+        // the correct item is selected (on right-click -> rename on not-current folder).
+        // TVN_ENDLABELEDIT becomes responsible for restoring the selection.
+    }
+    else
+    {
+        ++m_bNavigating;
+        TreeView_SelectItem(m_hWnd, m_oldSelected);
+        --m_bNavigating;
```

Comment:
```suggestion
        m_bNavigating = FALSE;
```
and same in the other instance.
Or chage this to a `bool` variable (see discussion in other PR) or to a numerical re-entry counter.

Reply:
We actually want ++ and -- here, I'll change away from BOOL instead.